lib/galaxy/model/__init__.py

```diff
@@ -3181,6 +3181,25 @@ def __create_version__(self, session):
             self.version = self.version + 1 if self.version else 1
             session.add(past_hda)
 
+    def copy_from(self, other_hda):
+        # This deletes the old dataset, so make sure to only call this on new things
+        # in the history (e.g. during job finishing).
+        old_dataset = self.dataset
+        self._metadata = None
+        self.metadata = other_hda.metadata or dict()
+        self.info = other_hda.info
+        self.blurb = other_hda.blurb
+        self.peek = other_hda.peek
+        self.extension = other_hda.extension
+        self.designation = other_hda.designation
+        self.deleted = other_hda.deleted
+        self.visible = other_hda.visible
+        self.validated_state = other_hda.validated_state
+        self.validated_state_message = other_hda.validated_state_message
+        self.copy_tags_from(self.history.user, other_hda)
```

Comment:
Should we set 
```suggestion
        self.copy_tags_from(self.history.user, other_hda)
        self.copied_from_history_dataset_association = other_hda.copied_from_history_dataset_association
```

?

Reply:
That is used for tracing back to a job during workflow extraction and this will have its own job though... Is there something else we use this for where we would need this tracking? If you think it is just good to track though I'm happy to add it... I think.

Reply:
Ah, I thought we used it for cached jobs, but we don't (and we don't need to). Let's not add it then.